### PR TITLE
Included tests approach

### PR DESCRIPTION
**Pull request by vlisianskii**
added new api to get related/disconnected/external tests filtered by unit/cucumber + included tests

### CODE REVIEW

File: engine/src/main/java/com/sandboni/core/engine/ResultGenerator.java [hunk 1]
```diff
@@ -119,6 +121,30 @@ private Object getContent(ResultContent rc) {
...
+                return graphOperations.getCucumberDisconnectedTests();
+            case RELATED_EXTERNAL_UNIT:
+                return graphOperations.getUnitRelatedExternalTests();
+            case RELATED_EXTERNAL_CUCUMBER:
+                return graphOperations.getCucumberRelatedExternalTests();
+            case ALL_EXTERNAL_UNIT:
+                return graphOperations.getAllExternalUnitTests();
+            case ALL_EXTERNAL_CUCUMBER:
+                return graphOperations.getAllExternalCucumberTests();
+            case ALL_TESTS_SIZE:
```
Comment (wmndev) [inline engine/src/main/java/com/sandboni/core/engine/ResultGenerator.java]: The 'size' entries (RELATED_TESTS_SIZE, ALL_TESTS_SIZE, DISCONNECTED_TESTS_SIZE)should not be here - whoever wants the size can take the returned list's size

File: engine/src/main/java/com/sandboni/core/engine/result/ResultContent.java [hunk 2]
```diff
@@ -7,13 +7,23 @@
...
+    FORMATTED_CHANGE_STATS(true),
+    RELATED_UNIT,
+    RELATED_CUCUMBER,
+    DISCONNECTED_UNIT,
+    DISCONNECTED_CUCUMBER,
+    RELATED_EXTERNAL_UNIT,
+    RELATED_EXTERNAL_CUCUMBER,
+    ALL_EXTERNAL_UNIT,
+    ALL_EXTERNAL_CUCUMBER,
+    ALL_TESTS_SIZE,
```
Comment (wmndev) [inline engine/src/main/java/com/sandboni/core/engine/result/ResultContent.java]: remove the *_SIZE

File: engine/src/main/java/com/sandboni/core/engine/sta/operation/VertexSizeOperation.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+import com.sandboni.core.engine.sta.graph.Edge;
+import com.sandboni.core.engine.sta.graph.vertex.TestVertex;
+import com.sandboni.core.engine.sta.graph.vertex.Vertex;
+import org.jgrapht.Graph;
+
+import java.util.Objects;
+import java.util.Set;
+
+public class VertexSizeOperation extends AbstractGraphOperation<LongResult> {
```
Comment (wmndev) [inline engine/src/main/java/com/sandboni/core/engine/sta/operation/VertexSizeOperation.java]: no need

File: engine/src/main/java/com/sandboni/core/engine/ResultGenerator.java (hunk 1 above)
Comment (vlisianskii) [reply engine/src/main/java/com/sandboni/core/engine/ResultGenerator.java]: I will remove RELATED_TESTS_SIZE and DISCONNECTED_TESTS_SIZE. But ALL_TESTS_SIZE should be here - maven plugin doesn't know about total test count in the project.

File: engine/src/main/java/com/sandboni/core/engine/sta/graph/vertex/CucumberVertex.java [hunk 4]
```diff
@@ -14,7 +14,7 @@
     @SuppressWarnings("squid:S00107") // private constructor used only in Builder
     private CucumberVertex(String actor, String action, boolean isSpecial, String filePath, List<Integer> lineNumbers, String filter,
                           String location, boolean ignore, boolean externalLocation, String featurePath, int scenarioLine, boolean affected) {
-        super(actor, action, isSpecial, filePath, lineNumbers, filter, location, ignore, externalLocation);
+        super(actor, action, isSpecial, filePath, lineNumbers, filter, location, ignore, externalLocation, false);
```
[the remark this answers is not in the document]
Comment (vlisianskii) [reply engine/src/main/java/com/sandboni/core/engine/sta/graph/vertex/CucumberVertex.java]: We cannot identify cucumber tests as "included"

File: engine/src/main/java/com/sandboni/core/engine/sta/operation/CucumberTestsOperation.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import java.util.stream.Collectors;
+
+public class CucumberTestsOperation extends AbstractGraphOperation<SetResult<CucumberVertex>> {
+    private final Set<TestVertex> tests;
+    private final boolean isExternal;
+
+    CucumberTestsOperation(Set<TestVertex> tests, boolean isExternal) {
+        Objects.requireNonNull(tests, INVALID_ARGUMENT);
+        this.tests = tests;
+        this.isExternal = isExternal;
```
[the remark this answers is not in the document]
Comment (vlisianskii) [reply engine/src/main/java/com/sandboni/core/engine/sta/operation/CucumberTestsOperation.java]: We have following API:
- related cucumber tests from current project
- related cucumber tests from external jar

For that one we have one CucumberTestsOperation with two parameters. No need to copy-paste the CucumberTestsOperation class for two operations.

File: engine/src/main/java/com/sandboni/core/engine/sta/operation/CucumberTestsOperation.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+import com.sandboni.core.engine.sta.graph.vertex.TestVertex;
+import com.sandboni.core.engine.sta.graph.vertex.Vertex;
+import org.jgrapht.Graph;
+
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class CucumberTestsOperation extends AbstractGraphOperation<SetResult<CucumberVertex>> {
+    private final Set<TestVertex> tests;
```
[the remark this answers is not in the document]
Comment (vlisianskii) [reply engine/src/main/java/com/sandboni/core/engine/sta/operation/CucumberTestsOperation.java]: Not sure. CucumberTestsOperation class is a filter. There could be "all tests", "related tests" or "external tests". That's why field called just "tests".

File: engine/src/main/java/com/sandboni/core/engine/sta/operation/UnitTestsOperation.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import com.sandboni.core.engine.sta.graph.vertex.TestVertex;
+import com.sandboni.core.engine.sta.graph.vertex.Vertex;
+import org.jgrapht.Graph;
+
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class UnitTestsOperation extends AbstractGraphOperation<SetResult<TestVertex>> {
+    private final Set<TestVertex> tests;
```
[the remark this answers is not in the document]
Comment (vlisianskii) [reply engine/src/main/java/com/sandboni/core/engine/sta/operation/UnitTestsOperation.java]: Not sure. UnitTestsOperation class is a filter. There could be "all tests", "related tests" or "external tests". That's why field called just "tests".

**Pull request merged**